ui/StatusQ/src/leftjoinmodel.cpp

```diff
@@ -0,0 +1,219 @@
+            && !m_rightModel->roleNames().empty())
+        initialize();
+}
+
+void LeftJoinModel::initialize()
+{
+    auto leftRoleNames = m_leftModel->roleNames();
+    auto rightRoleNames = m_rightModel->roleNames();
+
+    if (!leftRoleNames.count() ||  !rightRoleNames.count()) {
+        qWarning() << "Both left and right models have to contain some roles!";
+        return;
+    }
+
+    auto leftModelJoinRoleList = leftRoleNames.keys(m_joinRole.toUtf8());
+    auto rightModelJoinRoleList = rightRoleNames.keys(m_joinRole.toUtf8());
+
+    if (leftModelJoinRoleList.size() != 1
+            || rightModelJoinRoleList.size() != 1) {
+        qWarning().noquote() << QString("Both left and right models have to "
+                                        "contain join role %1!").arg(m_joinRole);
+        return;
+    }
+
+    m_leftModelJoinRole = leftModelJoinRoleList.at(0);
+    m_rightModelJoinRole = rightModelJoinRoleList.at(0);
+
+    auto leftRoles = leftRoleNames.keys();
+    auto maxLeftRole = std::max_element(leftRoles.cbegin(), leftRoles.cend());
+    auto rightRolesOffset = *maxLeftRole + 1;
+    auto roleNames = leftRoleNames;
+
+    auto i = rightRoleNames.constBegin();
+    while (i != rightRoleNames.constEnd()) {
+        if (i.value() != m_joinRole) {
+            auto roleWithOffset = i.key() + rightRolesOffset;
+            roleNames.insert(roleWithOffset, i.value());
+            m_joinedRoles.append(roleWithOffset);
+        }
+        ++i;
+    }
+
+    m_rightModelRolesOffset = rightRolesOffset;
+    m_roleNames = roleNames;
+
+    connect(m_rightModel, &QAbstractItemModel::dataChanged, this,
+            [this](auto& topLeft, auto& bottomRight, auto& roles) {
+        QVector<int> rolesTranslated;
+
+        if (roles.contains(m_rightModelJoinRole)) {
+            rolesTranslated = m_joinedRoles;
+        } else {
+            rolesTranslated = roles;
+
+            for (auto& role : rolesTranslated)
+                role += m_rightModelRolesOffset;
+        }
+
+        emit dataChanged(index(0, 0), index(rowCount() - 1, 0), rolesTranslated);
+    });
+
+    disconnect(m_leftModel, &QAbstractItemModel::rowsInserted,
+               this, &LeftJoinModel::initializeIfReady);
+    disconnect(m_rightModel, &QAbstractItemModel::rowsInserted,
+               this, &LeftJoinModel::initializeIfReady);
+
+    connect(m_rightModel, &QAbstractItemModel::rowsRemoved, this, [this] {
+        emit dataChanged(index(0, 0), index(rowCount() - 1, 0), m_joinedRoles);
+    });
+
+    connect(m_rightModel, &QAbstractItemModel::rowsInserted, this, [this] {
+        emit dataChanged(index(0, 0), index(rowCount() - 1, 0), m_joinedRoles);
+    });
+
+    connect(this, &QAbstractItemModel::dataChanged, this,
+            [this](auto& topLeft, auto& bottomRight, auto& roles ) {
+        if (roles.contains(m_leftModelJoinRole))
+            emit dataChanged(topLeft, bottomRight, m_joinedRoles);
+    });
```

Comment:
All these connections can be duplicated. We'd probably need unique connections here

Reply:
Now I see the `initialize` is guarded indirectly by ignoring model changes

Reply:
`dataChanged(const QModelIndex &topLeft, const QModelIndex &bottomRight, const QVector<int> &roles = QVector<int>())` -> missing the `const`s in the signature

Reply:
No, there is no need to mark `const` explicitly, it's const anyways, anyways it wouldn't compile (implicit constness removal?).

Quick evidence:

![Screenshot from 2023-10-26 17-50-14](https://github.com/status-im/status-desktop/assets/20650004/5e3719c6-8f58-4ee0-997a-ac82c78418ae)

Important here is only `&` to avoid copying.

Reply:
It compiles because the signature is normalized but yeah, not a big deal

Reply:
It compiles bc it's how auto works. Simplest example coming to my mind:
```
const int i = 0;
auto& j = i; // const int &
```
there is nothing special in our case with lambda